### PR TITLE
CLOUDP-283049: Fix Data Federation private endpoint reconciliation

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: internal/translation/datafederation/conversion_endpoints.go
Comment (josvazg) [inline internal/translation/datafederation/conversion_endpoints.go:74]: These are not equivalent, why the change?
Pointer conversions using the pointer package are more explicit and safe usually.

File: pkg/controller/atlasdatafederation/private_endpoint.go
Comment (josvazg) [inline pkg/controller/atlasdatafederation/private_endpoint.go:43]: This is the main fix right?

File: test/e2e/datafederation_pe_test.go
Comment (josvazg) [inline test/e2e/datafederation_pe_test.go:32]: Please remove focus here

**Review by josvazg — COMMENTED** [review]
LGTM over all, just one test needs focus removal and a few questions

**Review by josvazg — CHANGES_REQUESTED** [review]
Do not forget the test focus removal

**Comment by josvazg** [comment]
Separate feedback: This change is a fix that comes with a lot of refactors. While the refactors are welcome, most are probably not required for the fix and make reviewing it more difficult. My suggestion when this happens is to separate the refactor parts to a separate commit, even within the same PR. When doing this the PR should mention the separation so that reviewers can check each change as intended.

**Review by josvazg — APPROVED** [review]
LGTM (if CI agrees)

File: test/e2e/datafederation_pe_test.go
Comment (helderjs) [reply test/e2e/datafederation_pe_test.go]: mistake removed

File: internal/translation/datafederation/conversion_endpoints.go
Comment (helderjs) [reply internal/translation/datafederation/conversion_endpoints.go]: replaced back.

**Comment by helderjs** [comment]
> Separate feedback: This change is a fix that comes with a lot of refactors. While the refactors are welcome, most are probably not required for the fix and make reviewing it more difficult. My suggestion when this happens is to separate the refactor parts to a separate commit, even within the same PR. When doing this the PR should mention the separation so that reviewers can check each change as intended.

The refactor is the fix here. The goal was to not rely on the `Identifiable` interface and use `reflect.DeepEqual` instead. All the changes were to accommodate this approach and, of course, came with some improvements where possible.

**Comment by s-urbaniak** [comment]
Is only data federation affected?

File: internal/compare/comparation.go
Comment (s-urbaniak) [inline internal/compare/comparation.go:40]: this is not necessary, `reflect.DeepEquals` does what you are doing on slices.

**Comment by helderjs** [comment]
> Is only data federation affected?

No. I did it first because I thought it was the priority. Then I would check the others and create separate tickets

File: internal/compare/comparation.go
Comment (s-urbaniak) [reply internal/compare/comparation.go:40]: I simple `reflect.DeepEquals` should be enough, see https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1929/files#diff-11a090be55936627141c551ed7808292f8dc76db3baece47ad78ad24fb89d730R28

**Comment by s-urbaniak** [comment]
I just submitted https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1929 as well which is just missing some unit tests. I leave it up to you but I believe introducing a state machine is the way to go, wdyt?

File: internal/translation/datafederation/conversion_endpoints.go
Comment (s-urbaniak) [inline internal/translation/datafederation/conversion_endpoints.go:25]: I believe adding a map that considers the endpoint ID is more formally correct, see https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1929/files#diff-11a090be55936627141c551ed7808292f8dc76db3baece47ad78ad24fb89d730R73.

**Comment by helderjs** [comment]
Closing in favor of https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1929

**Pull request closed without merging**
